formulaic/transforms/basis_spline.py

```diff
@@ -98,18 +98,21 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
     if df is not None and knots is not None:
         raise ValueError("You cannot specify both `df` and `knots`.")
 
+    # Ensure ndarray to avoid conversion overhead
+    x = numpy.asarray(x)
+
     if "lower_bound" in _state:
-        lower_bound = _state["lower_bound"]
+        lower_bound = float(_state["lower_bound"])
```

Comment:
nit: Let's default to trusting what is in the `_state` dictionary. These are values that have already been prepared by a previous run through this code. And, even if they were numpy floats, it doesn't actually matter in practice. (also below)

Reply:
Nevermind... I realised that without it you have to add ugly typing annotations anyway. Ignore this one!